src/utils/flatten.js

```diff
@@ -14,28 +14,70 @@ export const objToCss = (obj: Object, prevKey?: string): string => {
 }` : css
 }
 
-const flatten = (chunks: Array<Interpolation>, executionContext: ?Object): Array<Interpolation> => (
-  chunks.reduce((ruleSet: Array<Interpolation>, chunk: ?Interpolation) => {
-    /* Remove falsey values */
-    if (chunk === undefined || chunk === null || chunk === false || chunk === '') return ruleSet
-    /* Flatten ruleSet */
-    if (Array.isArray(chunk)) return [...ruleSet, ...flatten(chunk, executionContext)]
-
-    /* Handle other components */
-    // $FlowFixMe not sure how to make this pass
-    if (chunk.hasOwnProperty('styledComponentId')) return [...ruleSet, `.${chunk.styledComponentId}`]
-
-    /* Either execute or defer the function */
-    if (typeof chunk === 'function') {
-      return executionContext
-        ? ruleSet.concat(...flatten([chunk(executionContext)], executionContext))
-        : ruleSet.concat(chunk)
+const innerFlatten = (
+  targetChunks: Array<Interpolation>,
+  chunk: Interpolation,
+  executionContext: ?Object,
+) => {
+  if (chunk === undefined || chunk === null || chunk === false || chunk === '') {
+    return
+  }
+
+  const chunkType = typeof chunk
+
+  if (chunkType === 'string') {
+    targetChunks.push(chunk)
+    return
+  }
+
+  if (chunkType === 'number') {
+    targetChunks.push(chunk.toString())
+    return
+  }
+
+  if (Array.isArray(chunk)) {
+    for (let i = 0; i < chunk.length; i += 1) {
+      innerFlatten(targetChunks, chunk[i], executionContext)
     }
+    return
+  }
+
+  if (chunkType === 'function') {
+    if (executionContext) {
+      // $FlowFixMe flow thinks chunk is a number
+      innerFlatten(targetChunks, chunk(executionContext), executionContext)
+    } else {
+      // defer the execution context, e.g. `css` styles
+      targetChunks.push(chunk)
+    }
+    return
+  }
+
+  if (chunk.hasOwnProperty('styledComponentId')) {
+    // $FlowFixMe flow thins chunk is a number here
```

Comment:
It might think so because of the typeof variable, maybe? Could break control flow inference